src/core/server/http/router/validator/validator.ts

```diff
@@ -0,0 +1,244 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ValidationError, Type, schema, ObjectType } from '@kbn/config-schema';
+import { Stream } from 'stream';
+import { RouteValidationError } from './validator_error';
+
+/**
+ * Allowed returned format of the custom validate function
+ * @public
+ */
+export type RouteValidateFunctionReturn<T> =
+  | {
+      value: T;
+      error?: never;
+    }
+  | {
+      value?: never;
+      error: RouteValidationError;
+    };
+
+/**
+ * The custom validation function if @kbn/config-schema is not a valid solution for your specific plugin requirements.
+ *
+ * @public
+ */
+export type RouteValidateFunction<T> = (data: any) => RouteValidateFunctionReturn<T>;
+
+/**
+ * Allowed property validation options: either @kbn/config-schema validations or custom validation functions
+ *
+ * @public
+ */
+export type RouteValidationSpec<T> = ObjectType | Type<T> | RouteValidateFunction<T>;
+
+// Ugly as hell but we need this conditional typing to have proper type inference
+type RouteValidatedValue<T extends RouteValidationSpec<any> | undefined> = NonNullable<
+  T extends RouteValidateFunction<any>
+    ? ReturnType<T>['value']
+    : T extends Type<any>
+    ? ReturnType<T['validate']>
+    : undefined
+>;
+
+/**
+ * The configuration object to the RouteValidator class.
+ * Set `params`, `query` and/or `body` to specify the validation logic to follow for that property.
+ *
+ * @public
+ */
+export interface RouteValidatorConfig<P, Q, B> {
+  /**
+   * Validation logic for the URL params
+   * @public
+   */
+  params?: RouteValidationSpec<P>;
+  /**
+   * Validation logic for the Query params
+   * @public
+   */
+  query?: RouteValidationSpec<Q>;
+  /**
+   * Validation logic for the body payload
+   * @public
+   */
+  body?: RouteValidationSpec<B>;
+}
+
+/**
+ * Additional options for the RouteValidator class to modify its default behaviour.
+ *
+ * @public
+ */
+export interface RouteValidatorOptions {
+  /**
+   * Set the `unsafe` config to avoid running some additional internal *safe* validations on top of your custom validation
+   * @public
+   */
+  unsafe?: {
+    params?: boolean;
+    query?: boolean;
+    body?: boolean;
+  };
+}
+
+/**
+ * Route validations config and options merged into one object
+ * @public
+ */
+export type RouteValidatorFullConfig<P, Q, B> = RouteValidatorConfig<P, Q, B> &
+  RouteValidatorOptions;
+
+/**
+ * Route validator class to define the validation logic for each new route.
+ *
+ * @private
```

Comment:
I'm not sure about the behavior differences, but we've been using `@internal`

Reply:
My bad! Well spotted!